### PR TITLE
Delete IGMP Groups when deleting stale chassis.

### PR DESCRIPTION
**Pull request by npinaeva**
IGMPGroup has a weak link to chassis, deleting multiple chassis may result in IGMP_Groups with identical values on columns "address", "datapath", and "chassis", when "chassis" goes empty

Signed-off-by: Nadia Pinaeva <npinaeva@redhat.com>

Error message example:
```
error:constraint violation Details:Transaction causes multiple rows in "IGMP_Group" table to have identical values 
(mrouters, 038b16fa-6aba-4244-9d4f-00a1e2cbf9a2, and []) for index on columns "address", "datapath", and "chassis".  
First row, with UUID 7e9a18fa-e58c-4547-a7cb-afa934b6cdc9, had the following index values before the transaction: 
mrouters, 038b16fa-6aba-4244-9d4f-00a1e2cbf9a2, and d9755997-e909-4d0c-8770-82a902d69a90.  Second row, with 
UUID 84da3622-3ac7-41f0-a6b5-536a2d5f9137, had the following index values before the transaction: mrouters, 
038b16fa-6aba-4244-9d4f-00a1e2cbf9a2, and 578d4dd9-cc02-4bcc-8a9c-08dcc3a94190. UUID:{GoUUID:} Rows:[]}] 
```

### CODE REVIEW

**Review by flavio-fernandes — COMMENTED** [review]
nice!
/lgtm

File: go-controller/pkg/libovsdb/libovsdb.go [hunk 1]
```diff
@@ -102,6 +102,7 @@ func NewSBClientWithConfig(cfg config.OvnAuthConfig, promRegistry prometheus.Reg
 
 	// Only Monitor Required SBDB tables to reduce memory overhead
 	chassisPrivate := sbdb.ChassisPrivate{}
+	igmpGroup := sbdb.IGMPGroup{}
```
Comment (flavio-fernandes) [inline go-controller/pkg/libovsdb/libovsdb.go:105]: Note: there is a cost associated to monitoring a table. This may affect our scaling numbers. ☣︎

**Review by trozet — APPROVED** [review]

**Comment by trozet** [comment]
CI for control-plane failed to launch kind

**Comment by trozet** [comment]
The problem is when multicast is enabled, ovnkube-master crashes:
```
I0202 20:18:52.165623      41 cache.go:1007] cache "msg"="creating row" "database"="OVN_Southbound" "model"="&{UUID:30a18308-144b-4287-8f23-1a2020df36ad Address: Chassis:0xc0005c88b0 Datapath:<nil> Ports:[]}" "table"="IGMP_Group" "uuid"="30a18308-144b-4287-8f23-1a2020df36ad"
panic: gob: cannot encode nil pointer of type *string

goroutine 1 [running]:
encoding/gob.(*Encoder).EncodeValue(0xc0003bc280?, {0x1a11900?, 0x0?, 0x0?})
	/usr/lib/golang/src/encoding/gob/encoder.go:223 +0x679
encoding/gob.(*Encoder).Encode(0x0?, {0x1a11900?, 0x0?})
	/usr/lib/golang/src/encoding/gob/encoder.go:176 +0xa5
github.com/ovn-org/libovsdb/cache.valueFromIndex(0x1b85800?, {0xc0004fec00, 0x3, 0x18?})
	/home/trozet/go/src/github.com/ovn-kubernetes/go-controller/vendor/github.com/ovn-org/libovsdb/cache/cache.go:1436 +0x2ec
github.com/ovn-org/libovsdb/cache.(*RowCache).Create(0xc000eef880, {0xc00039c780, 0x24}, {0x1c43620?, 0xc000622460}, 0x0)
```

This happens because gob cannot encode nil values or nil pointers. IGMP_Group is indexed by:
"Contains learned IGMP groups indexed by address/datapath/chassis."

Datapath and chassis are string pointers:
```
type IGMPGroup struct {
	UUID     string   `ovsdb:"_uuid"`
	Address  string   `ovsdb:"address"`
	Chassis  *string  `ovsdb:"chassis"`
	Datapath *string  `ovsdb:"datapath"`
	Ports    []string `ovsdb:"ports"`
}
```

and in the above log we can see datapath is nil. It looks like this code in libovsdb is trying to create some unique encoded value to serve as an index. If I simply skip trying to encode values that are nil, KIND works:

```
--- a/go-controller/vendor/github.com/ovn-org/libovsdb/cache/cache.go
+++ b/go-controller/vendor/github.com/ovn-org/libovsdb/cache/cache.go
@@ -1433,6 +1433,15 @@ func valueFromIndex(info *mapper.Info, columnKeys []model.ColumnKey) (interface{
                        if err != nil {
                                return "", err
                        }
+                       // if object is nil dont try to encode it
+                       value := reflect.ValueOf(val)
+                       if value.Kind() == reflect.Invalid {
+                               continue
+                       }
+                       // if object is a nil pointer dont try to encode it
+                       if value.Kind() == reflect.Pointer && value.IsNil() {
+                               continue
+                       }
                        err = enc.Encode(val)
                        if err != nil {
                                return "", err

```

I'm not sure this is the right solution though as then it might make the encoded value non unique if we skip a key? Not sure. @dave-tucker @jcaamano maybe you know.

**Comment by dave-tucker** [comment]
> I'm not sure this is the right solution though as then it might make the encoded value non unique if we skip a key? Not sure. @dave-tucker @jcaamano maybe you know.

Your root cause is correct and also the fix is fine too. This was an assumption that I'd made which was "index values will always be non-nil" which unfortunately we've found out is not true.

**Comment by trozet** [comment]
Thanks @dave-tucker . Pushed https://github.com/ovn-org/libovsdb/pull/351

**Comment by flavio-fernandes** [comment]
@trozet @npinaeva : I started a PR to bump libovsdb. Please feel free to move that into this PR, if it helps:
https://github.com/ovn-org/ovn-kubernetes/pull/3404

**Review by trozet — APPROVED** [review]

**Pull request merged**